crates/iceberg/src/catalog/rest.rs

```diff
@@ -0,0 +1,912 @@
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module contains rest catalog implementation.
+
+use std::collections::HashMap;
+
+use async_trait::async_trait;
+use reqwest::header::{self, HeaderMap, HeaderName, HeaderValue};
+use reqwest::{Client, Request};
+use serde::de::DeserializeOwned;
+use urlencoding::encode;
+
+use crate::error::Result;
+use crate::table::Table;
+use crate::{
+    Catalog, Error, ErrorKind, Namespace, NamespaceIdent, TableCommit, TableCreation, TableIdent,
+};
+
+use self::_serde::{
+    CatalogConfig, ErrorModel, ErrorResponse, ListNamespaceResponse, ListTableResponse,
+    NamespaceSerde, RenameTableRequest, NO_CONTENT, OK,
+};
+
+const ICEBERG_REST_SPEC_VERSION: &str = "0.14.1";
+const PATH_V1: &str = "v1";
+
+#[derive(Debug, Builder)]
+pub struct RestCatalogConfig {
+    uri: String,
+    #[builder(default)]
+    warehouse: Option<String>,
+
+    #[builder(default)]
+    props: HashMap<String, String>,
+}
+
+impl RestCatalogConfig {
+    fn config_endpoint(&self) -> String {
+        [&self.uri, PATH_V1, "config"].join("/")
+    }
+
+    fn namespaces_endpoint(&self) -> String {
+        [&self.uri, PATH_V1, "namespaces"].join("/")
+    }
+
+    fn namespace_endpoint(&self, ns: &NamespaceIdent) -> Result<String> {
+        Ok([&self.uri, PATH_V1, "namespaces", &ns.encode_in_url()?].join("/"))
+    }
+
+    fn tables_endpoint(&self, ns: &NamespaceIdent) -> Result<String> {
+        Ok([
+            &self.uri,
+            PATH_V1,
+            "namespaces",
+            &ns.encode_in_url()?,
+            "tables",
+        ]
+        .join("/"))
+    }
+
+    fn rename_table_endpoint(&self) -> Result<String> {
+        Ok([&self.uri, PATH_V1, "tables", "rename"].join("/"))
+    }
+
+    fn table_endpoint(&self, table: &TableIdent) -> Result<String> {
+        Ok([
+            &self.uri,
+            PATH_V1,
+            "namespaces",
+            &table.namespace.encode_in_url()?,
+            "tables",
+            encode(&table.name).as_ref(),
+        ]
+        .join("/"))
+    }
+
+    fn try_create_rest_client(&self) -> Result<HttpClient> {
+        //TODO: We will add oauth, ssl config, sigv4 later
+        let mut headers = HeaderMap::new();
+        headers.insert(
+            header::CONTENT_TYPE,
+            HeaderValue::from_static("application/json"),
+        );
+        headers.insert(
+            HeaderName::from_static("x-client-version"),
+            HeaderValue::from_static(ICEBERG_REST_SPEC_VERSION),
+        );
+        headers.insert(
+            header::USER_AGENT,
+            HeaderValue::from_str(&format!("iceberg-rs/{}", env!("CARGO_PKG_VERSION"))).unwrap(),
+        );
+
+        Ok(HttpClient(
+            Client::builder().default_headers(headers).build()?,
+        ))
+    }
+}
+
+impl NamespaceIdent {
+    /// Returns url encoded format.
+    pub fn encode_in_url(&self) -> Result<String> {
+        if self.0.is_empty() {
```

Comment:
It's better to ensure that `NamespaceIdent` is valid so that we don't have to check it when using it. This change can removes a lot of `Result<T>` in related APIs.